go/ct/common/hash256.go

```diff
@@ -0,0 +1,100 @@
+//
+// Copyright (c) 2024 Fantom Foundation
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at fantom.foundation/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use
+// of this software will be governed by the GNU Lesser General Public Licence v3
+//
+
+package common
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/Fantom-foundation/Tosca/go/vm"
+	"pgregory.net/rand"
+)
+
+// Hash256 is an immutable slice of Hash256 that can be trivially cloned.
+type Hash256 struct {
+	data [256]vm.Hash
```

Comment:
Objects of this struct may be immutable, but it is not cheap to compare or copy.

The memory layout of this type is 256 vm.Hash values next to each other. When you compare them, all 8 KiB of data is compared, and when you assign them 8 KiB of data is copied.

To make it light-weight, the object should contain a pointer to the data. This pointer can be trivially cloned, restored, and in comparisons you can first check the pointer to the data. If it is the same, there is no need to compare 8 KiB. If it is different, you need to compare the data. But the idea is to have it unlikely that in this case the data is actually equivalent, and the comparison will abort after a few bytes.

Reply:
changes applied, benchmarks rerun.